File: frontend/src/lib/readme-media-resolve.ts
import { parseGithubRepoUrl } from "@/lib/github-url"

function normalizeRepoPath(path: string): string {
  const parts = path.split("/").filter(Boolean)
  const out: string[] = []
  for (const part of parts) {
    if (part === ".") continue
    if (part === "..") {
      out.pop()
      continue
    }
    out.push(part)
  }
  return out.join("/")
}

/** README 所在目录（不含文件名），默认仓库根。 */
export function readmeDirectoryFromBasePath(readmeBasePath: string | null | undefined): string {
  if (!readmeBasePath?.trim()) return ""
  const parts = normalizeRepoPath(readmeBasePath.trim()).split("/").filter(Boolean)
  if (parts.length <= 1) return ""
  parts.pop()
  return parts.join("/")
}

/** GitHub raw 内容 base URL，用于解析相对图片/资源路径。 */
export function readmeRawBaseUrl(
  githubUrl: string | undefined,
  readmeBasePath?: string | null
): string | null {
  if (!githubUrl?.trim()) return null
  const parsed = parseGithubRepoUrl(githubUrl)
  if (!parsed) return null
  const dir = readmeDirectoryFromBasePath(readmeBasePath)
  const suffix = dir ? `${dir}/` : ""
  return `https://raw.githubusercontent.com/${parsed.full_name}/HEAD/${suffix}`
}

/** 将 README 内图片 src 解析为可加载的绝对 URL。 */
export function resolveReadmeImageSrc(
  src: string | undefined,
  githubUrl: string | undefined,
  readmeBasePath?: string | null
): string | undefined {
  if (!src?.trim()) return src
  const trimmed = src.trim()
  if (
    trimmed.startsWith("http://") ||
    trimmed.startsWith("https://") ||
    trimmed.startsWith("data:") ||
    trimmed.startsWith("//")
  ) {
    return trimmed
  }
  const base = readmeRawBaseUrl(githubUrl, readmeBasePath)
  if (!base) return trimmed
  try {
    return new URL(trimmed, base).href
  } catch {
    return trimmed
  }
}

/** 解析 `<source srcset="...">` 中的相对 URL（含 1x/2x 描述符）。 */
export function resolveReadmeSrcSet(
  srcSet: string | undefined,
  githubUrl: string | undefined,
  readmeBasePath?: string | null
): string | undefined {
  if (!srcSet?.trim()) return srcSet
  return srcSet
    .split(",")
    .map((candidate) => {
      const trimmed = candidate.trim()
      if (!trimmed) return trimmed
      const parts = trimmed.split(/\s+/)
      const url = parts[0] ?? ""
      const resolved = resolveReadmeImageSrc(url, githubUrl, readmeBasePath) ?? url
      if (parts.length <= 1) return resolved
      return `${resolved} ${parts.slice(1).join(" ")}`
    })
    .join(", ")
}

File: backend/app/services/readme_media_resolve.py
"""README 内相对图片路径解析（对齐前端 readme-media-resolve.ts）。"""

from __future__ import annotations

import re
from urllib.parse import urljoin

from app.services.github_parse import parse_owner_repo


def _normalize_repo_path(path: str) -> str:
    parts = [p for p in path.split("/") if p and p != "."]
    out: list[str] = []
    for part in parts:
        if part == "..":
            if out:
                out.pop()
            continue
        out.append(part)
    return "/".join(out)


def readme_directory_from_base_path(readme_base_path: str | None) -> str:
    if not readme_base_path or not readme_base_path.strip():
        return ""
    parts = _normalize_repo_path(readme_base_path.strip()).split("/")
    if len(parts) <= 1:
        return ""
    return "/".join(parts[:-1])


def github_normalized_url(full_name: str, github_url: str | None) -> str | None:
    parsed = parse_owner_repo(full_name, github_url)
    if not parsed:
        return None
    owner, repo = parsed
    return f"https://github.com/{owner}/{repo}"


def readme_raw_base_url(full_name: str, github_url: str | None, readme_base_path: str | None) -> str | None:
    base = github_normalized_url(full_name, github_url)
    if not base:
        return None
    directory = readme_directory_from_base_path(readme_base_path)
    suffix = f"{directory}/" if directory else ""
    return f"{base}/raw/HEAD/{suffix}"


def resolve_readme_image_src(
    src: str | None,
    *,
    full_name: str,
    github_url: str | None,
    readme_base_path: str | None,
) -> str | None:
    if not src or not src.strip():
        return src
    trimmed = src.strip()
    if re.match(r"^(https?:|data:|//)", trimmed, re.I):
        return trimmed
    base = readme_raw_base_url(full_name, github_url, readme_base_path)
    if not base:
        return trimmed
    return urljoin(base, trimmed)
